chat-plugins/mafia.js

```diff
@@ -0,0 +1,610 @@
+		this.currentVote = {};
+		let noMafia = {};
+		let mafia = [];
+
+		for (let i in this.players) {
+			let player = this.players[i];
+
+			if (player.class.side !== 'mafia') {
+				noMafia[i] = player;
+			} else {
+				mafia.push(player);
+			}
+		}
+
+		for (let i = 0; i < mafia.length; i++) {
+			mafia[i].voting = true;
+			mafia[i].validVotes = noMafia;
+
+			let flavorText = '';
+			if (mafia.length === 1) {
+				flavorText += 'As the only live member of the mafia, you have to be careful. Not careful enough to stop killing, though.';
+			} else if (mafia.length === 2) {
+				flavorText += 'You sit down with the only other member of the mafia, ' + (i === 0 ? Tools.escapeHTML(mafia[1].name) : Tools.escapeHTML(mafia[0].name)) + '.';
+			} else {
+				flavorText += 'You sit down with the other members of the mafia, ';
+				for (let j = 0; i < mafia.length; i++) {
+					if (i !== j) {
+						if (j === (mafia.length - 1) || (j < i && j === (mafia.length - 2))) {
+							flavorText += ' and ';
+						} else {
+							flavorText += ', ';
+						}
+						flavorText += Tools.escapeHTML(mafia[i].name);
+					}
+				}
+			}
+
+			mafia[i].voteWindow(mafia[i].class.image, flavorText);
+		}
+	}
+
+	townMeeting() {
+		this.meeting = 'mafia';
+		this.currentVote = {};
+
+		for (let i in this.players) {
+			let player = this.players[i];
+
+			if (this.currentVote[player.userid]) {
+				this.currentVote[player.userid]++;
+			} else {
+				this.currentVote[player.userid] = 1;
+			}
+			player.voting = true;
+			player.validVotes = this.players;
+
+			player.voteWindow(player.class.image, 'Outraged over the mafia\'s activity in town, the people decide to lynch a person they suspect of being involved with the mafia.');
+		}
+	}
+
+	gameEvent(gamestate, event, timer) {
+		this.gamestate = gamestate;
+
+		this.executionOrder = [];
+
+		for (let i in this.players) {
+			let player = this.players[i];
+			if (player.class[event]) {
+				player.event(event);
+			}
+		}
+
+		this.executionOrder.sort(function (a, b) {
+			return (b.class[event].priority - a.class[event].priority);
+		});
+
+		this.setTimer(timer);
+	}
+}
+
+exports.commands = {
+	mafia: {
+		create: 'new',
+		new: function (target, room, user) {
+			let params = target.split(',').map(function (param) { return param.toLowerCase().trim(); });
+
+			if (!this.can(permission, null, room)) return false;
+			if (!room.mafiaEnabled) return this.errorReply("Mafia is disabled for this room.");
+			if (!this.canTalk()) return this.errorReply("You cannot do this while unable to talk.");
+			if (room.game) return this.errorReply("There is already a game in progress in this room.");
+
+			// TODO: make a generator for a default setup.
+			if (!params) return this.errorReply("No roles entered.");
+
+			for (let i = 0; i < params.length; i++) {
+				if (!(params[i] in MafiaData.MafiaClasses)) {
+					return this.errorReply(Tools.escapeHTML(params[i]) + " is not a valid mafia class.");
+				}
+			}
```

Comment:
This seems to be a very cumbersome syntax

``/mafia new town, town, town, town, mafia, mafia,...``

It should probably be something like ``/mafia new 10 town, 8 mafia, 1 doctor``, etc